man/man3/acl_to_any_text.3
.\" Access Control Lists manual pages
.\"
.\" (C) 2002 Andreas Gruenbacher, <a.gruenbacher@computer.org>
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.TH ACL_TO_ANY_TEXT 3 "Linux ACL Library" "March 2002" "Access Control Lists"
.SH NAME
acl_to_any_text \- convert an ACL to text
.SH LIBRARY
Linux Access Control Lists library (libacl, \-lacl).
.SH SYNOPSIS
.sp
.nf
.B #include <sys/types.h>
.B #include <sys/acl.h>
.B #include <acl/libacl.h>
.sp
.B "char *acl_to_any_text (acl_t \f2acl\f3, const char *\f2prefix\f3, "
.B "                       char \f2separator\f3, int \f2options\f3);"
.Op
.SH DESCRIPTION
The
.B acl_to_any_text
function translates the ACL pointed to by the argument
.I acl
into a NULL terminated character string. This character string
is composed of the ACL entries contained in
.IR acl ,
in the entry text format described on
.BR acl (5).
Entries are separated from each other by the
.I separator
character.
If the argument
.I prefix
is not NULL,
each entry is prefixed by this character string.
.PP
If the argument
.I options
is 0, ACL entries are converted using the entry tag type keywords
.BR user ,
.BR group ,
.BR mask ,
and
.BR other .
User IDs and group IDs of ACL entries that contain such
qualifiers are converted to their corresponding names; if an identifier
has no corresponding name, a decimal number string is produced. The
ACL text representation contains no additional comments.
.PP
A bitwise combinations of the following
.I options
can be used to modify the result:
.TP
.B TEXT_ABBREVIATE
Instead of the full tag type keywords, single letter abbreviations are used.
The abbreviation for
.B user
is
.BR u ,
the abbreviation for
.B group
is
.BR g ,
the abbreviation for
.B mask
is
.BR m ,
and the abbreviation for
.B other
is
.BR o .
.TP
.B TEXT_NUMERIC_IDS
User IDs and group IDs are included as decimal numbers instead of names.
.TP
.B TEXT_SOME_EFFECTIVE
A comment containing the effective permissions of the ACL entry is
included after ACL entries that contain permissions which are ineffective
because they are masked by an ACL_MASK entry. The ACL entry and the comment
are separated by a tab character.
.TP
.B TEXT_ALL_EFFECTIVE
A comment containing the effective permissions of the ACL entry is
included after all ACL entries that are affected by an ACL_MASK entry.
The comment is included even if the permissions contained in the ACL
entry equal the effective permissions. The ACL entry and the comment are
separated by a tab character.
.TP
.B TEXT_SMART_INDENT
This option is used in combination with the TEXT_SOME_EFFECTIVE or
TEXT_ALL_EFFECTIVE option. The number of tab characters inserted between
the ACL entry and the comment is increased so that the comment is
aligned to at least column 32, if printed starting from column 0. A tab
width of 8 characters is assumed.
.PP
The ACL referred to by
.I acl
is not changed.
.PP
This function allocates any memory necessary to contain the string and
returns a pointer to the string.  The caller should free any releasable
memory, when the new string is no longer required, by calling
.BR acl_free (3)
with the char* returned by
.B acl_to_any_text
as an argument.
.SH RETURN VALUE
Upon successful completion, this function returns a pointer to the text
representation of the ACL.  Otherwise, a value of NULL is returned, and
.B errno
is set to indicate the error.
.SH ERRORS
If any of the following conditions occur, the
.B acl_to_any_text
function returns a value of NULL and sets
.B errno
to the corresponding value:
.TP
.SM
\%[EINVAL]
The argument
.I acl
is not a valid pointer to an ACL.
.TP
.SM
\%[EINVAL]
The ACL referenced by
.I acl
contains one or more improperly formed ACL entries, or for some other
reason cannot be translated into the text form of an ACL.
.TP
.SM
\%[ENOMEM]
The character string to be returned requires more memory than is allowed
by the hardware or system-imposed memory management constraints.
.SH STANDARDS
This is a non-portable, Linux specific extension to the ACL manipulation
functions defined in IEEE Std 1003.1e draft 17 (\(lqPOSIX.1e\(rq, abandoned).
.SH AUTHOR
Written by
.I "Andreas Gruenbacher"
<a.gruenbacher@computer.org>.
.SH SEE ALSO
.BR acl_to_text (3),
.BR acl_free (3),
.BR acl (5)


doc/CHANGES
2.0.6 (08 April 2002)
* Fix a bug that causes setfacl not to report error messages if
  acl-set_file() fails (the bug was introduced in version 2.0.3).

2.0.5 (26 March 2002)
* Fix up acl_to_text compliance with POSIX draft.
* Add the new acl(5) manual page.
* Add a manual section 3 directory and Makefile.
* Add the libacl manual pages.
* Rename the existing acl_to_any_text function to __acl_to_any_text, and
  add a new acl_to_any_text function that only contains parameters that
  are actually useful.  Make acl_to_text and acl_to_any_text invoke
  __acl_to_any_text internally.  Update all uses of acl_to_any_text in
  getfacl/getfacl.c, setfacl/do_set.c, and chacl/chacl.c.
* Move the acl_size function form the LIBACL_CFILES to the POSIX_CFILES
  section (it is a POSIX function)
* Rearrange libacl routines so that several non-POSIX routines which
  were being exported are no longer exported (eg. acl_entry_to_any_str,
  acl_get_file_mode, acl_get_fd_mode, acl_set_file_mode, acl_set_fd_mode).

2.0.4 (12 March 2002)
* Man page updates from Andreas
* Test script updates from Andreas

2.0.3 (05 March 2002)
* Clean up the --default option handling in setfacl. The old
  workarounds caused a bug for unusual input.
* Changes to the --test output format setfacl generates: ACLs that
  are not changed are now displayed as `*'.
* Fix a bug in setfacl/sequence.c:seq_delete_cmd().
* Minor changes to test scripts

2.0.2 (01 March 2002)
* Apply several patches from Andreas, namely:
* man page fixes
* libacl code reformatting
* acl_from_text errno handling

2.0.1 (27 February 2002)
* Apply Andreas Gruenbacher's diffs.
* Fix up chacl for deletion of access ACL to be in line with Andreas. 
* Incorporate the Debian packaging again.

2.0.0 (21 February 2002)
* Reworked to use the new official system call API.
* Sync up with the XFS project, the SGI folk now use this source.
* Jumped to version 2 to allow XFS users to upgrade
  (Rationale: the XFS ACL user tools were at version 1.1.X, and
  packaging tools like rpm, dpkg, etc. must be presented with a
  greater version number to allow an upgrade to proceed).
* Added the chacl command to ease migration for existing XFS users,
  and for compatibility with IRIX.
* Added a flag to allow acl_print to produce a single-line ACL, in
  addition to the multi-line format.
* Extended attribute documentation has moved into the extended
  attribute package from SGI ("attr"), this ACL package now deals
  exclusively with ACLs.
* acl_from_text sometimes did not set errno when failing.
* Moved files and simplified #includes in libacl

0.7.22
* Create a shared version of libacl, instead of linking statically.
* Added missing error reporting to getfacl do_print() function.     

0.7.16
* Minor cleanups/fixes in the make files
* The test cases in test/src have been updated to match the
  corrected ACL entry syntax (see the 0.7.15 changes).

0.7.15
* acl_from_text() now accepts mask and other entries with either two
  or three fields (`other:rwx' or `other::rwx'). Posix 1003.1e specifies
  that these entries should have three entries, while under Solaris,
  they have only two fields.
* setfacl now parses mask and other entries with two or three entries.
* getfacl now produces mask and other entries with three fields
  (`mask::rwx' instead of the previous `mask:rwx').
* acl_get_fd(), acl_get_file(), acl_set_fd(), acl_set_file() now only
  fail with errno=ENOSYS if the underlying filesystem doesn't support
  ACLs. acl_get_fd() and acl_get_file() construct ACLs from the file
  mode permission bits if the kernel supports ACLs, but no ACL extended
  attributes are defined for the file (errno=ENOATTR).
* acl_to_text() wrongly separated ACL entires with ',' instead of '\n'.
* acl_from_text() only accepted three-letter permission strings
  (rwx, ---, rw-, etc.). Now also short strings (rw, -, etc.) are
  supported.

0.7.10
* acl_get_fd(), acl_get_file() now don't return basic ACL's anymore.
* acl_set_fd(), acl_set_file() fail on filesystems that don't support
  ACLs, even if the ACL to be set is a basice ACL.
* If msgfmt is not present on a system, the message catalog(s) are
  no longer being built.

0.7.8 (19 February 2001)
* Buffer overflow bug in ACL library (acl_to_any_text).
  (Reported by Charles Bertsch <CBertsch@microtest.com>)

0.7.2 (22 November 2000)
* Replace rman with groff for building HTML versions of the
  manual pages.
* Add descriptions of the permissions required for manipulating
  ACLs to the manual pages.
* Add long options to getfacl and setfacl, update the manual pages.
* Remove some small bugs from the src/showacl Perl script.
* showacl is now integrated in getfacl.
* Bug in acl_to_any_text() for ACLs with 0 entries fixed.
* Bug in acl_to_text(): Used '\n' as the ACL entry separator.
* German messages updated.

0.7.1 (23 October 2000)
* Basic GNU gettext support and a German translation of messages
  added.

0.7.0 (22 October 2000)
* Updates to the documentation
* Now an ACL_MASK entry is cloned from the ACL_GROUP_OBJ entry
  when needed also when the ACL_MASK entry is not re-calculated
  (-n option), so setfacl doesn't complain about an invalid ACL.

0.6.8 (11 October 2000)
* src/walk_tree.c was stat()ing the targets of symbolic links in
  WALK_PHYSICAL mode.
* Removed spurious definitions from src/gettext.h that caused
  problems with several versions of GNU gettext headers
* setfacl calls acl_delete_def_file() if removing the default ACL
  of a directory that doesn't have a default ACL. i(If the user lacks
  appropriate permissions, that no-operation fails.) Fixed.
* Tiny bug in src/showacl script (mask entry not applied to owning
  group entry)

0.6.5 (10 September 2000)
* Command line options revised for POSIX compatibility (now defaults to
  -H half-logical walk, added -L logical walk, -P physical walk).
* Funny post-order option no longer documented.

0.6.0 (24 March 2000)
* Permissions were not set to the union of the ACL_GROUP_OBJ and

libacl/acl_to_any_text.c
/*
  File: acl_to_any_text.c

  Copyright (C) 1999, 2000
  Andreas Gruenbacher, <a.gruenbacher@computer.org>

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU Library General Public
  License as published by the Free Software Foundation; either
  version 2 of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Library General Public License for more details.

  You should have received a copy of the GNU Library General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

#include <stdio.h>
#include <errno.h>
#include <acl/libacl.h>
#include "libacl.h"

char *
acl_to_any_text(
	acl_t acl,
	ssize_t *len_p,
	const char *prefix,
	char separator,
	const char *suffix,
	int options)
{
	acl_obj *acl_obj_p = ext2int(acl, acl);
	ssize_t size, len = 0, entry_len = 0,
		suffix_len = suffix ? strlen(suffix) : 0;
	string_obj *string_obj_p, *tmp;
	acl_entry_obj *entry_obj_p, *mask_obj_p = NULL;
	if (!acl_obj_p)
		return NULL;
	size = acl->a_used * 15 + 1;
	string_obj_p = new_var_obj_p(string, size);
	if (!string_obj_p)
		return NULL;

	if (options & (TEXT_SOME_EFFECTIVE|TEXT_ALL_EFFECTIVE)) {
		/* fetch the ACL_MASK entry */
		FOREACH_ACL_ENTRY(entry_obj_p, acl_obj_p) {
			if (entry_obj_p->etag == ACL_MASK) {
				mask_obj_p = entry_obj_p;
				break;
			}
		}
	}

	FOREACH_ACL_ENTRY(entry_obj_p, acl_obj_p) {
		if (len + entry_len + 1 > size) {
			while (len + entry_len + 1 > size)
				size <<= 1;
			tmp = realloc_var_obj_p(string, string_obj_p, size);
			if (tmp == NULL)
				goto fail;
			string_obj_p = tmp;
		}

		entry_len = acl_entry_to_any_str(int2ext(entry_obj_p),
		                                 string_obj_p->sstr + len,
						 size-len,
						 int2ext(mask_obj_p),
						 prefix,
						 options);
		if (entry_len < 0)
			goto fail;
		if (len + entry_len + suffix_len + 1 > size)
			continue;
		len += entry_len;
		string_obj_p->sstr[len] = separator;
		len++;
	}
	if (len)
		len--;
	if (len && suffix) {
		strcpy(string_obj_p->sstr + len, suffix);
		len += suffix_len;
	} else
		string_obj_p->sstr[len] = '\0';

	if (len_p)
		*len_p = len;
	return (char *)int2ext(string_obj_p);

fail:
	free_obj_p(string_obj_p);
	return NULL;
}

